File: Data Structures & Algorithms/top-k-elements-in-list/submission-2.js
class Solution {
    topKFrequent(nums, k) {
      // Step 1: count
    const countMap = new Map();
    for (const num of nums) {
        countMap.set(num, (countMap.get(num) || 0) + 1);
    }

    // Step 2: sort entries by frequency descending
    const sorted = [...countMap.entries()].sort((a, b) => b[1] - a[1]);
    //  countMap.entries() gives: [[1,3], [2,2], [3,1]]
    //  after sort:               [[1,3], [2,2], [3,1]]  (already sorted here)
    //
    //  sort logic: b[1] - a[1]
    //  b[1] = frequency of b
    //  a[1] = frequency of a
    //  if result > 0, b comes first (descending)

    // Step 3: take top k, extract just the number (not frequency)
    return sorted.slice(0, k).map(entry => entry[0]);
}
}

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int len = nums.length;
        int[] result = new int[len];
        int[] leftProd = new int[len];
        int[] rightProd = new int[len];
        if(len<=1){
            return nums;
        }
         leftProd[0] = nums[0];
         rightProd[len-1] = nums[len-1];
        for(int i=1; i<len; i++){
            leftProd[i] = leftProd[i-1] * nums[i];
        }
        for(int i=len-2; i>=0; i--){
            rightProd[i] = rightProd[i+1] * nums[i];
        }

        result[0]= rightProd[1];
        result[len-1]= leftProd[len-2];
        for(int i=1; i<len-1; i++){
            result[i] = leftProd[i-1]*rightProd[i+1];
        }
        return result;
    }
}  
